app/logins.py

```diff
@@ -0,0 +1,425 @@
+    active_flow = request.session.get("active-login-flow", None)
+    if active_flow is not None:
+        ret["method"] = active_flow
+        ret["method_intermediate"] = request.session["active-login-flow-intermediate"]
+        if ret["state"] == LoginState.LOGGED_IN:
+            ret["state"] = LoginState.LOGGING_IN_AGAIN
+        else:
+            ret["state"] = LoginState.LOGGING_IN
+    return ret
+
+
+@router.get("/login")
+def get_login_kinds():
+    """
+    Retrieve the login methods available from the backend.
+
+    For each method returned, flow starts with a `GET` to the endpoint
+    `.../login/{method}` and upon completion from the user-agent, with a `POST`
+    to that same endpoint name.
+
+    Each method is also given a button icon and some text to use, though
+    frontends with localisation may choose to render other text instead.
+    """
+    return [
+        {
+            "method": "github",
+            "button": "https://login-with-github-icon.png",
+            "text": "Log in with Github",
+        }
+    ]
+
+
+@router.get("/login/github")
+def start_github_flow(request: Request, login=Depends(login_state)):
+    """
+    Starts a github login flow.  This will set session cookie values and
+    will return a redirect.  The frontend is expected to save the cookie
+    for use later, and follow the redirect to Github
+
+    Upon return from Github to the frontend, the frontend should POST to this
+    endpoint with the relevant data from Github
+
+    If the user is already logged in, and has a valid github token stored,
+    then this will return an error instead.
+    """
+    if login["state"].logging_in():
+        # Already logging in to something
+        if login["method"] == "github":
+            # Already logging into github, so assume something went squiffy
+            # and send them back with the same in-progress login
+            pass
+        else:
+            # Already mid-flow to something else
+            return JSONResponse(
+                {"state": "error", "error": "User is mid-login-flow elsewhere"},
+                status_code=400,
+            )
+    user = login["user"]
+    if user is not None:
+        gha = models.GithubAccount.by_user(db, user)
+        if gha is not None and gha.token is not None:
+            return JSONResponse(
+                {"state": "error", "error": "User already logged into github"},
+                status_code=400,
+            )
+    # Okay, we're preparing a login, step one, do we already have an
+    # intermediate we're using?
+    models.GithubFlowToken.housekeeping(db)
+    intermediate = login["method_intermediate"]
+    if intermediate is not None:
+        # Yes, retrieve it from the db
+        intermediate = db.session.get(models.GithubFlowToken, intermediate)
+        if intermediate:
+            created = intermediate.created
+            now = datetime.now()
+            if (now - created).seconds > (10 * 60):
```

Comment:
I decided to split the difference since the user might loiter at the github login for a bit - tokens now default to housekeeping at 20 minutes, and that's the only way they expire.